gherkin/php/src/TokenMatcher.php

```diff
@@ -0,0 +1,229 @@
+        }
+
+        return false;
+    }
+
+    public function match_FeatureLine(Token $token): bool
+    {
+        return $this->matchTitleLine($token, TokenType::FeatureLine, $this->currentDialect->getFeatureKeywords());
+    }
+
+    public function match_BackgroundLine(Token $token): bool
+    {
+        return $this->matchTitleLine($token, TokenType::BackgroundLine, $this->currentDialect->getBackgroundKeywords());
+    }
+
+    public function match_ScenarioLine(Token $token): bool
+    {
+        return $this->matchTitleLine($token, TokenType::ScenarioLine, $this->currentDialect->getScenarioKeywords())
+            || $this->matchTitleLine($token, TokenType::ScenarioLine, $this->currentDialect->getScenarioOutlineKeywords());
+    }
+
+    public function match_RuleLine(Token $token): bool
+    {
+        return $this->matchTitleLine($token, TokenType::RuleLine, $this->currentDialect->getRuleKeywords());
+    }
+
+    public function match_ExamplesLine(Token $token): bool
+    {
+        return $this->matchTitleLine($token, TokenType::ExamplesLine, $this->currentDialect->getExamplesKeywords());
+    }
+
+    /**
+     * @param list<non-empty-string> $keywords
+     */
+    private function matchTitleLine(Token $token, TokenType $tokenType, array $keywords): bool
+    {
+        foreach ($keywords as $keyword) {
+            if ($token->line?->startsWithTitleKeyword($keyword)) {
+                $title = $token->line->getRestTrimmed(StringUtils::symbolCount($keyword) + StringUtils::symbolCount(GherkinLanguageConstants::TITLE_KEYWORD_SEPARATOR));
+                $this->setTokenMatched($token, $tokenType, $title, $keyword);
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public function match_Other(Token $token): bool
+    {
+        //take the entire line, except removing DocString indents
+        $text = $token->line?->getLineText($this->indentToRemove) ?? '';
+        $this->setTokenMatched($token, TokenType::Other, $this->unescapeDocString($text), indent: 0);
+
+        return true;
+    }
+
+    public function match_Empty(Token $token): bool
+    {
+        if ($token->line?->isEmpty()) {
+            $this->setTokenMatched($token, TokenType::Empty);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    public function match_StepLine(Token $token): bool
+    {
+        $keywords = $this->currentDialect->getStepKeywords();
+        foreach ($keywords as $keyword) {
+            if ($token->line?->startsWith($keyword)) {
+                $stepText = $token->line->getRestTrimmed(StringUtils::symbolCount($keyword));
+                $this->setTokenMatched($token, TokenType::StepLine, $stepText, $keyword);
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public function match_TableRow(Token $token): bool
+    {
+        if ($token->line?->startsWith(GherkinLanguageConstants::TABLE_CELL_SEPARATOR)) {
```

Comment:
might be worth inverting the condition here?

Reply:
Maybe later; I'd want to change it in the Java version too and don't have the confidence yet!